src/v/config/configuration.cc

```diff
@@ -1414,6 +1414,12 @@ configuration::configuration()
       "Manifest upload timeout (ms)",
       {.needs_restart = needs_restart::no, .visibility = visibility::tunable},
       10s)
+  , cloud_storage_garbage_collect_timeout_ms(
+      *this,
+      "cloud_storage_garbage_collect_timeout_ms",
+      "Timeout for running the cloud storage garbage collection (ms)",
+      {.needs_restart = needs_restart::no, .visibility = visibility::tunable},
+      30s)
```

Comment:
suggestion: This makes sense to me. We know the cloud storage backend type at start-up. Perhaps we can configure some default values for each type at start-up? This can be follow-up though.

Reply:
@VladLazar I'd like to think about this setting as being independent of the backend. Think of it rather as a _budget_ for how long the archival service can spend on the cleanup before falling to the orphaning behavior (`cloud_storage_max_segments_pending_deletion_per_partition`).

For the per-backend setting, I'd suggest we introduce a new setting `cloud_storage_api_operation_timeout_ms` which we apply in the `remote` class for each api call (less so for the blob upload/download). If backend supports batch operations, then the timeout will be for the entire batch call. If the backend doesn't support batch, then the timeout will be per operation.

I believe this to be a better way for achieving fault tolerance and resiliency. The per-api timeout ensures we don't get stuck with bad connections. The gc timeout ensures we don't accumulate too much in-memory metadata (protects the system to run out of memory) if we fail to make progress for a prolonged duration.

Thoughts?